service/s3/api_client.go

```diff
@@ -259,6 +259,22 @@ func disableAcceptEncodingGzip(stack *middleware.Stack) error {
 	return acceptencodingcust.AddAcceptEncodingGzip(stack, acceptencodingcust.AddAcceptEncodingGzipOptions{})
 }
 
+// ResponseError provides the HTTP centric error type wrapping the underlying error
+// with the HTTP response value and the deserialized RequestID.
+type ResponseError interface {
+	error
+
+	ServiceHostID() string
+	ServiceRequestID() string
```

Comment:
Nit: Just curious if the following could have been an option to expose this. 

Mostly thinking about additions to metadata would break interface. Though that's rare and probably can just introduce an extra interface in that case.  So this option can be ruled out but still like to know your thoughts. 

```
type ResponseError interface {
   error
    
   ServiceMetadata(key interface{}) (interface{}, error)
}

key being S3.HostIDKey{} or aws.RequestIDKey{} or any other metadata type added in future for S3. 
```

This adds the overhead of specifying a key type and return of a loose value. This does pair up with how context works though.

Reply:
ServiceResponseID is a member of the `aws/transport/http#ResponseError` so in order to expand on that  method the specific method would be needed. This is a part of the core difference between errors and responses. Errors can be layered with interfaces matching specific layers and metadata. Whereas output cannot be layered and only option is via metadata accessors.

This does make error and response slightly asymmetric, but the purpose fits in their common use cases better.  If only metadata were used on the errors, it wouldn't be possible to use strong typing to layer errors the way that is being done between, smithy's ResponseError, aws's ResponseError, and S3's ResponseError.